lib/WebService/Async/Segment.pm

```diff
@@ -0,0 +1,251 @@
+
+    return $self->{ua} if $self->{ua};
+
+    $self->{ua} = Net::Async::HTTP->new(
+        fail_on_error            => 1,
+        decode_content           => 1,
+        pipeline                 => 0,
+        stall_timeout            => 60,
+        max_connections_per_host => 2,
+        user_agent               => 'Mozilla/4.0 (WebService::Async::Segment; BINARY@cpan.org; https://metacpan.org/pod/WebService::Async::Segment)',
+    );
+
+    $self->add_child($self->{ua});
+
+    return $self->{ua};
+}
+
+=head2 basic_authentication
+
+Settings required for basic HTTP authentication
+
+=cut
+
+sub basic_authentication {
+    my $self = shift;
+
+    #C<Net::Async::Http> basic authentication information
+    return {
+        user => $self->write_key,
+        pass => ''
+    };
+}
+
+=head2 absolute_uri
+
+Constructs the absolute URI based on two params:
+
+=over 4
+
+=item * C<relative_uri> - relative address of a server resource (API method name)
+
+=item * C<args> - HTTP args to be included in the uri
+
+=back
+
+=cut
+
+sub absolute_uri {
+    my ($self, $relative_uri, %args) = @_;
+    URI::Template->new($self->base_uri . $relative_uri)->process(%args);
+}
+
+=head2 method_call
+
+Makes a Segment method call based on params. It automatically sets B<sentAt> to current time and B<context->{library}> to the current module.
+
+It takes two params:
+
+=over 4
+
+=item * C<method> - required. Segment method name (such as B<identify> and B<track>).
+
+=item * C<args> - optional. Method arguments represented as a hash. It may include either common, method-specific or custom fields.
+Please refer to L<https://segment.com/docs/spec/common/> for a full list of common fieds supported by Segment.
+
+=back
+
+=cut
+
+sub method_call {
+    my ($self, $method, %args) = @_;
+
+    $args{sentAt}                        = Date::Utility->new()->datetime_iso8601;
+    $args{context}->{library}->{name}    = ref $self;
+    $args{context}->{library}->{version} = $VERSION;
+
+    return Future->fail('Method cannot be empty') unless $method;
+
+    return Future->fail('Both userId and anonymousId are empty') unless $args{userId} or $args{anonymousId};
+
+    $log->tracef('Segment method %s called with params %s', $method, \%args);
+
+    return $self->ua->POST(
+        $self->absolute_uri($method),
+        encode_json_utf8(\%args),
+        content_type => 'application/json',
+        %{$self->basic_authentication},
+        )->then(
+        sub {
+            my $result = shift;
+
+            $log->tracef('Segment response for %s method received: %s', $method, $result);
+
+            try {
+                my $response = decode_json_utf8($result->content);
+                if ($response->{success}) {
+                    $log->tracef('Segment %s method call finished successfully.', $method);
+
+                    return Future->done($response->{success});
+                }
+                return Future->fail($response);
+            }
```

Comment:
For failures, note that it's a 3-part system: the exception message or object, the category, then any details. for example `Future->fail('InvalidServerResponse', 'segment', $@, $response)` https://metacpan.org/pod/Future#fail1

Reply:
Changed accordingly. Thanks.